### PR TITLE
Introduced engines abstraction layer

### PR DESCRIPTION
**Pull request by favreau**
In order to be able to choose between OSPRay, Optix and Firerays as acceleration libraries, an abstraction of the ray-tracing engine is needed. This pull requests implements those new engine classes, as well as the factory that is to be used to instanciate them. The engine can be chosen using the --engine command line option.

### CODE REVIEW

File: brayns/Brayns.cpp [hunk 1]
```diff
@@ -46,72 +44,44 @@ namespace brayns
...
+        if( !_engine )
+            throw std::runtime_error( "Unsupported engine: " + engineName );
 
         // set HDRI skybox if applicable
         const std::string& hdri =
             _parametersManager->getRenderingParameters().getHDRI();
+        ScenePtr scene = _engine->getScene();
         if( !hdri.empty() )
-            _scene->getMaterial(MATERIAL_SKYBOX)->setTexture(TT_DIFFUSE, hdri);
+            scene->getMaterial(MATERIAL_SKYBOX)->setTexture(TT_DIFFUSE, hdri);
```
Comment (bilgili) [inline brayns/Brayns.cpp]: this is very engine specific ?

Comment (bilgili) [inline brayns/Brayns.cpp]: Different engines may have different initializations.

File: brayns/Brayns.cpp [hunk 2]
```diff
@@ -46,72 +44,44 @@ namespace brayns
...
+        scene->addLight( sunLight );
 
+        // Build geometry
         loadData( );
-        _scene->buildEnvironment( );
-        _scene->buildGeometry( );
-        _scene->commit( );
+        scene->buildEnvironment( );
+        scene->buildGeometry( );
+        scene->commit( );
```
Comment (bilgili) [inline brayns/Brayns.cpp]: commit is also OSPray centric.

File: brayns/Brayns.cpp [hunk 3]
```diff
@@ -137,16 +107,15 @@ struct Brayns::Impl
...
     }
 
     void render( const RenderInput& renderInput,
                  RenderOutput& renderOutput )
     {
         reshape( renderInput.windowSize );
 
-        _frameBuffer->map( );
-        _camera->set(
+        _engine->getFrameBuffer()->map( );
```
Comment (bilgili) [inline brayns/Brayns.cpp]: map is OSPray centric. I think the initialization should be in the engine construction.

File: brayns/Brayns.cpp (hunk 1 above)
Comment (favreau) [inline brayns/Brayns.cpp]: No, this is not engine specific. The texture types (diffuse, normal, etc) are generic, as well as the ID for the material ID for the Skybox. It's the job of the engine to set the texture accordingly for the renderer.

File: brayns/Brayns.cpp (hunk 2 above)
Comment (favreau) [inline brayns/Brayns.cpp]: Well, not really. GPU engine such as Optix also need to transfer data from CPU to GPU memory and this is the job of the commit method. That will eventually also be useful for out-of-core rendering.

File: brayns/Brayns.cpp (hunk 3 above)
Comment (favreau) [inline brayns/Brayns.cpp]: I agree. Changing it. Thanks for that! :)

File: brayns/Brayns.cpp [hunk 4]
```diff
@@ -267,47 +242,46 @@ struct Brayns::Impl
...
         const Vector3f diag   = worldBounds.getSize( );
         Vector3f position = target;
         position.z( ) -= diag.z( );
 
         const Vector3f up  = Vector3f(0.f,1.f,0.f);
-        _camera->setInitialState(position, target, up);
-        _camera->setAspectRatio(
-            static_cast< float >( _frameSize.x()) /
-            static_cast< float >( _frameSize.y()));
+        camera->setInitialState(position, target, up);
```
Comment (bilgili) [inline brayns/Brayns.cpp]: looks like copy pasted code. normalize it !

File: plugins/engines/Engine.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "Engine.h"
+
+namespace brayns
+{
+
+void Engine::setActiveRenderer( const std::string renderer )
```
Comment (bilgili) [inline plugins/engines/Engine.cpp]: pass by ref string

File: plugins/extensions/plugins/DeflectPlugin.cpp [hunk 6]
```diff
@@ -66,9 +67,11 @@ void DeflectPlugin::_initializeDeflect()
 
 void DeflectPlugin::_sendDeflectFrame()
 {
-    const Vector2i frameSize = _extensionParameters.frameBuffer->getSize();
+    FrameBufferPtr frameBuffer =
+        _extensionParameters.engine->getFrameBuffer();
+    const Vector2i frameSize = frameBuffer->getSize();
```
Comment (bilgili) [inline plugins/extensions/plugins/DeflectPlugin.cpp]: const ref

File: plugins/engines/Engine.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+       Commits changes to the engine. This include scene and camera
+       modifications
+    */
+    virtual void commit() = 0;
+
+    /** Renders the current scene and populates the frame buffer accordingly */
+    virtual void render() = 0;
+
+    /** Executes engine specific pre-render operations */
+    virtual void preRender() = 0;
```
Comment (bilgili) [inline plugins/engines/Engine.h]: I think post and pre functions does not need to be pure. You can add an empty implementation. Not every engine may need them.

File: brayns/Brayns.cpp (hunk 4 above)
Comment (favreau) [inline brayns/Brayns.cpp]: Not sure if I understand what you mean there :-/ Can you explain a bit more?

File: plugins/engines/Engine.cpp (hunk 5 above)
Comment (favreau) [inline plugins/engines/Engine.cpp]: Indeed! Shame on me.

File: plugins/engines/Engine.h (hunk 7 above)
Comment (favreau) [inline plugins/engines/Engine.h]: Hum... The idea of forcing the implementation in the inherited classes is one way to remind the developper that those methods exist, and prevent him/her to put the code in the render() method instead. That's why I'd rather keep them pure, but I'd like to have your opinion on that :-)

File: plugins/extensions/plugins/DeflectPlugin.cpp (hunk 6 above)
Comment (favreau) [inline plugins/extensions/plugins/DeflectPlugin.cpp]: Yep... shame on me again :'(

File: brayns/Brayns.cpp (hunk 4 above)
Comment (bilgili) [inline brayns/Brayns.cpp]: There are multiple places the same code, can you gather into a function. Normalization means decreasing the repetition.

File: plugins/engines/Engine.h (hunk 7 above)
Comment (bilgili) [inline plugins/engines/Engine.h]: Developer should be aware of that. They are generally going to have a look at headers for implementation or example implementations.

**Comment by bilgili** [comment]
+1

File: plugins/engines/Engine.h (hunk 7 above)
Comment (favreau) [inline plugins/engines/Engine.h]: Fair enough

**Pull request merged**